File: waverless_ui/src/apis.ts
import axios from "axios"


export class NodeBasic {
    constructor(
        public name:string,
        public online:boolean,
        public ip:string,
        public ssh_port:string,
        public cpu_sum:number,
        public cpu_cur:number,
        public mem_sum:number,
        public mem_cur:number,
        public passwd:string,
        public system:string,
    ){}
}

export class Action {
    constructor(
        public name:string,
        public cmd:string,
    ){}
}

export class ServiceBasic {
    constructor(
        public name:string,
        public node:string,
        public dir:string,
        public actions:Action[],
    ){}
}


export class AddServiceRespSucc {
    constructor(

    ){}
}

export class AddServiceRespTemplate {
    constructor(
        public nodes:string[],
    ){}
}

export class AddServiceRespFail {
    constructor(
        public msg:string,
    ){}
}

export class AddServiceResp{
    constructor(
        private kernel: any,
        private id: number
    ) {}
    
    succ():undefined| AddServiceRespSucc{
        if(this.id==1){
            return this.kernel
        }
        return undefined
    }
    
    template():undefined| AddServiceRespTemplate{
        if(this.id==2){
            return this.kernel
        }
        return undefined
    }
    
    fail():undefined| AddServiceRespFail{
        if(this.id==3){
            return this.kernel
        }
        return undefined
    }
    
}


export class AddServiceReq {
    constructor(
        public service:ServiceBasic,
    ){}
}

export namespace apis {
    export async function add_service(req:AddServiceReq):Promise<AddServiceResp>{
        let res:any = await axios.post("/api/add_service", req)
        return new AddServiceResp(res.data.kernel,res.data.id)
    }
}




export class DeleteServiceRespSucc {
    constructor(

    ){}
}

export class DeleteServiceRespFail {
    constructor(
        public msg:string,
    ){}
}

export class DeleteServiceResp{
    constructor(
        private kernel: any,
        private id: number
    ) {}
    
    succ():undefined| DeleteServiceRespSucc{
        if(this.id==1){
            return this.kernel
        }
        return undefined
    }
    
    fail():undefined| DeleteServiceRespFail{
        if(this.id==2){
            return this.kernel
        }
        return undefined
    }
    
}


export class DeleteServiceReq {
    constructor(
        public service:string,
    ){}
}

export namespace apis {
    export async function delete_service(req:DeleteServiceReq):Promise<DeleteServiceResp>{
        let res:any = await axios.post("/api/delete_service", req)
        return new DeleteServiceResp(res.data.kernel,res.data.id)
    }
}




export class GetServiceListRespExist {
    constructor(
        public services:ServiceBasic[],
    ){}
}

export class GetServiceListResp{
    constructor(
        private kernel: any,
        private id: number
    ) {}
    
    exist():undefined| GetServiceListRespExist{
        if(this.id==1){
            return this.kernel
        }
        return undefined
    }
    
}


export namespace apis {
    export async function get_service_list():Promise<GetServiceListResp>{
        let res:any = await axios.post("/api/get_service_list", )
        return new GetServiceListResp(res.data.kernel,res.data.id)
    }
}




export class RunServiceActionRespSucc {
    constructor(
        public output:string,
    ){}
}

export class RunServiceActionRespFail {
    constructor(
        public msg:string,
    ){}
}

export class RunServiceActionResp{
    constructor(
        private kernel: any,
        private id: number
    ) {}
    
    succ():undefined| RunServiceActionRespSucc{
        if(this.id==1){
            return this.kernel
        }
        return undefined
    }
    
    fail():undefined| RunServiceActionRespFail{
        if(this.id==2){
            return this.kernel
        }
        return undefined
    }
    
}


export class RunServiceActionReq {
    constructor(
        public service:string,
        public action_cmd:string,
        public sync:boolean,
    ){}
}

export namespace apis {
    export async function run_service_action(req:RunServiceActionReq):Promise<RunServiceActionResp>{
        let res:any = await axios.post("/api/run_service_action", req)
        return new RunServiceActionResp(res.data.kernel,res.data.id)
    }
}

File: src/main/src/apis.rs
use crate::general::network::http_handler::ApiHandlerImpl;
use async_trait::async_trait;
use axum::{http::StatusCode, routing::post, Json, Router};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use crate::metrics::metrics_handler;
use axum::routing::get;

#[derive(Debug, Serialize, Deserialize)]
pub struct NodeBasic {
    pub name: String,
    pub online: bool,
    pub ip: String,
    pub ssh_port: String,
    pub cpu_sum: f64,
    pub cpu_cur: f64,
    pub mem_sum: f64,
    pub mem_cur: f64,
    pub passwd: String,
    pub system: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Action {
    pub name: String,
    pub cmd: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ServiceBasic {
    pub name: String,
    pub node: String,
    pub dir: String,
    pub actions: Vec<Action>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum AddServiceResp {
    Succ {},
    Template { nodes: Vec<String> },
    Fail { msg: String },
}

impl AddServiceResp {
    fn id(&self) -> u32 {
        match self {
            AddServiceResp::Succ { .. } => 1,
            AddServiceResp::Template { .. } => 2,
            AddServiceResp::Fail { .. } => 3,
        }
    }
    pub fn serialize(&self) -> Value {
        json!({
            "id": self.id(),
            "kernel": serde_json::to_value(self).unwrap(),
        })
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AddServiceReq {
    pub service: ServiceBasic,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum DeleteServiceResp {
    Succ {},
    Fail { msg: String },
}

impl DeleteServiceResp {
    fn id(&self) -> u32 {
        match self {
            DeleteServiceResp::Succ { .. } => 1,
            DeleteServiceResp::Fail { .. } => 2,
        }
    }
    pub fn serialize(&self) -> Value {
        json!({
            "id": self.id(),
            "kernel": serde_json::to_value(self).unwrap(),
        })
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DeleteServiceReq {
    pub service: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum GetServiceListResp {
    Exist { services: Vec<ServiceBasic> },
}

impl GetServiceListResp {
    fn id(&self) -> u32 {
        match self {
            GetServiceListResp::Exist { .. } => 1,
        }
    }
    pub fn serialize(&self) -> Value {
        json!({
            "id": self.id(),
            "kernel": serde_json::to_value(self).unwrap(),
        })
    }
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum RunServiceActionResp {
    Succ { output: String },
    Fail { msg: String },
}

impl RunServiceActionResp {
    fn id(&self) -> u32 {
        match self {
            RunServiceActionResp::Succ { .. } => 1,
            RunServiceActionResp::Fail { .. } => 2,
        }
    }
    pub fn serialize(&self) -> Value {
        json!({
            "id": self.id(),
            "kernel": serde_json::to_value(self).unwrap(),
        })
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RunServiceActionReq {
    pub service: String,
    pub action_cmd: String,
    pub sync: bool,
}

#[async_trait]
pub trait ApiHandler {
    async fn handle_add_service(&self, req: AddServiceReq) -> AddServiceResp;

    async fn handle_delete_service(&self, req: DeleteServiceReq) -> DeleteServiceResp;

    async fn handle_get_service_list(&self) -> GetServiceListResp;

    async fn handle_run_service_action(&self, req: RunServiceActionReq) -> RunServiceActionResp;
}

pub fn add_routers(mut router: Router) -> Router {
    async fn add_service(Json(req): Json<AddServiceReq>) -> (StatusCode, Json<Value>) {
        (
            StatusCode::OK,
            Json(ApiHandlerImpl.handle_add_service(req).await.serialize()),
        )
    }
    router = router.route("/add_service", post(add_service));

    async fn delete_service(Json(req): Json<DeleteServiceReq>) -> (StatusCode, Json<Value>) {
        (
            StatusCode::OK,
            Json(ApiHandlerImpl.handle_delete_service(req).await.serialize()),
        )
    }
    router = router.route("/delete_service", post(delete_service));

    async fn get_service_list() -> (StatusCode, Json<Value>) {
        (
            StatusCode::OK,
            Json(ApiHandlerImpl.handle_get_service_list().await.serialize()),
        )
    }
    router = router.route("/get_service_list", post(get_service_list));

    async fn run_service_action(Json(req): Json<RunServiceActionReq>) -> (StatusCode, Json<Value>) {
        (
            StatusCode::OK,
            Json(
                ApiHandlerImpl
                    .handle_run_service_action(req)
                    .await
                    .serialize(),
            ),
        )
    }
    router = router.route("/run_service_action", post(run_service_action));
    router
}
